kvbc/include/categorization/kv_blockchain.h

```diff
@@ -179,6 +180,13 @@ class KeyValueBlockchain {
   // currently we are operating with single thread
   util::ThreadPool thread_pool_{1};
 
+  // metrics
+  std::shared_ptr<concordMetrics::Aggregator> aggregator_;
+  concordMetrics::Component metrics_;
+  concordMetrics::CounterHandle versioned_num_of_deletes_keys_;
+  concordMetrics::CounterHandle immuteable_num_of_deleted_keys_;
+  concordMetrics::CounterHandle merkle_num_of_deleted_keys_;
+
```

Comment:
What does it mean to use metrics in the storage layer? I remember that in order to publish them some timer needs to lock and push the updates right? does it mean that we'll introduce locks to the flow? 
Also using those counters which are atomics have significant cost.
Personally I more and more tend to approach using log-based metrics.

Reply:
To address your concerns, I changed the component to be a delete component only.
It will update the aggregator every time we call the deleteBlock method.
Notice that update the aggregator requires grabbing a lock (as you mentioned), however, since we delete blocks only on pruning it doesn't really matter.
The counters themselves are not atomic, but again, if they were, we call delete block only on pruning.
I agree that we should consider enhancing our metrics platform (or maybe take a different approach) but currently, this is what we have

Reply:
cool

Reply:
@yontyon  is correct, a lock is only taken when aggregating stats. This is pretty much the case for all efficient metrics systems, including HdrHistogram. 

The key design is to make updating stats as fast as possible as it happens very frequently, and in the hot path most of the time. Therefore all counters are just local increments using non-atomic variables. They shouldn't take much more than a single cpu instruction, although I haven't checked the assembly. On the other hand, when an observer needs to view the collected metrics, they must be aggregated to a central location. This currently requires taking a lock on that central location so that the data can be copied to the central location. The only contention on that lock is from other writers who are updating simultaneously. It is this point where we may be able to change such that we don't require a lock, although in all cases some sort of synchronization will be required.

While much of the code updates the aggregator with a timer, because this is a straightforward mechanism to keep the centralized stats reasonably up to date, aggregation does not have to be done that way. As Yehonatan did here, it's just a matter of calling the `UpdateAggregator` method. 

I strongly think that all metrics should still be updates to local variables and only be made atomic when they must be updatable from multiple threads, as is the case now. The thing that we may want to change is the aggregation itself. One thing that @toly-kournik  and I discussed was getting rid of the aggregator altogether and instead using a `pull` mechanism, rather than a `push` mechanism. We would make the metrics register a callable and then have the diagnostics code (or similar) execute the callable on demand. In practice, running this callable would either require taking a lock or publishing to a promise (which may take a lock under the hood).  It's doubtful this would be much more efficient than the current code, but it would have the benefit that arbitrary timers wouldn't be used to publish metrics. The code would only pull when needed. So instead of publishing every 100ms even if stats are only updated on a dashboard every 5s, we'd pull once every 5s rather than 50 times. This also has the upside, that a single central component won't face much, if any contention for locks, and those locks will be spread across components (granular). The downside of this strategy is that now some central process decides when to take a lock on each independent component, rather than that independent component deciding when to publish when its most efficient for it to do so. We don't want a lock being taken in a hot path for storage and then blocking while the stats are copied out to the shared data structure. It also may be harder to structure the code such that you can update freely in a local variable and only take a lock an publish, as now two threads are accessing that data rather than just one.

Reply:
I simply addressed Eran's comment. I think that for now, our metrics system balances the tradeoffs well.

Reply:
Yeah I know. I liked your explanation, but since this has come up a few times I figured I'd add a bit more detail for the record. I actually do think it's worthwhile to consider changing the metrics code as we rework our architecture and it's worth discussing in the future.